Assets/Scripts/PackItUp/Managers/GameStateManager.cs

```diff
@@ -0,0 +1,117 @@
+using System;
+using System.Collections.Generic;
+using PackItUp.Controllers;
+using PackItUp.Interactables;
+using PackItUp.MockSystems;
+using UnityEngine;
+
+namespace PackItUp.Managers {
+    public class GameStateManager : MonoBehaviour
+    {
+        // Inform that the player completed the level successfully
+        public event EventHandler OnLevelSuccess;
+
+        // Inform that the player failed to complete the level
+        public event EventHandler OnLevelFailed;
+
+        // Inform of a game start
+        public event EventHandler OnLevelStart;
+
+
+        // Inform that the player completed the objective
+        public event EventHandler OnObjectiveCompleted;
+
+        // There will only be one game manager that handles when to start a new game loop
+        private GameManager _gameManager;
+
+        // Win condition will start as false as the player has not completed the objective of the game
+        private bool _winCondition = false;
+
+        // We need to know how many players are present in the game so we can enable/disable its movement between levels
+        private List<TopDownCharacterController> _players;
+
+        // A single inventory will be present in the game, this is managed by the game manager 
+        //and we should listen to the event the inventory emits when all key items are collected
+        private MockInventory _inventory;
+
+        // The timer exists only in the game manager, we need to listen to the event emmited when the timer runs out
+        private MockTimer _timer;
+
+        // The current level, this will live in the game manager, the game state manager is agnostic of which level is the player on
+        private MockLevel _currentLevel;
+
+        // In each level, we'll have an end zone, this will be activate once the objective is completed allowing the player 
+        // to end the game in a success state
+        private EndZone[] _endZones;
+
+        private void Awake() 
+        {
+            _gameManager = GameManager.Instance;
+            _timer = _gameManager.GetTimer();
+            _inventory = _gameManager.GetInventory();
+            _players = _gameManager.GetPlayers();
+            _currentLevel = _gameManager.GetLevel();
+            _endZones = _currentLevel.GetEndZones();
+        }
+
+        private void OnEnable()
+        {
+            _gameManager.OnGameStart += StartGame;
+            _timer.OnTimerRunOut += EndGameFailedState;
+            EndZone.OnPlayerEnteredZone += TryEndGameSuccessfully;
+            EndZone.OnPlayerExitZone += CancelEndGameCountdown;
+            _inventory.OnKeyItemsCollected += CompleteObjective;
+        }
+
+        private void OnDisable()
+        {
+            _gameManager.OnGameStart -= StartGame;
+            _timer.OnTimerRunOut -= EndGameFailedState;
+            EndZone.OnPlayerEnteredZone -= TryEndGameSuccessfully;
+            EndZone.OnPlayerExitZone += TryEndGameSuccessfully;
```

Comment:
Trying to end the game both on entry and on exit might be overkill, let's say we have a single player, then the remaining items cannot be retrieved inside the end zone. If we have more players, it might make sense as we'll check whether the other player got the remaining items.

However I feel that it's a bit of a weird game experience,right? To _exit_ the end zone to trigger a check, maybe we can have it constantly triggering or do some workarounds like this:

- If the player enters the end zone, notify the game state manager that will update a flag accordingly
- If the player exits, the same, update that boolean
- If the inventory gets completed *and* that flag is true (Meaning that we have other player inside the exit zone) then we can proceed


With this we avoid constantly sending events from the end zone and constantly checking for the players 👀